utils/linux/proc_events.py

```diff
@@ -0,0 +1,214 @@
+        """Runs forever and calls registered callbacks on process events"""
+        try:
+            self._socket.bind((os.getpid(), self._CN_IDX_PROC))
+        except PermissionError as e:
+            raise PermissionError("You don't have permissions to bind to the process events connector") from e
+
+        self._register_for_connector_events(self._socket)
+        self._selector.register(self._socket, selectors.EVENT_READ)
+
+        while not self._should_stop:
+            events = self._selector.select()
+            if self._should_stop:
+                break
+
+            for key, _ in events:
+                data = key.fileobj.recv(256)
+
+                nl_hdr = dict(
+                    zip(("len", "type", "flags", "seq", "pid"), self._nlmsghdr.unpack(data[: self._nlmsghdr.size]))
+                )
+                if nl_hdr["type"] != self._NLMSG_DONE:
+                    # Handle only netlink messages
+                    continue
+
+                # Strip off headers
+                data = data[self._nlmsghdr.size : nl_hdr["len"]]
+                data = data[self._cn_msg.size :]
+
+                event = dict(
+                    zip(
+                        ("what", "cpu", "timestamp_ns"),
+                        self._base_proc_event.unpack(data[: self._base_proc_event.size]),
+                    )
+                )
+
+                if event["what"] == self._PROC_EVENT_EXIT:
+                    # (Notice that exit_signal is the signal that the parent process received on exit, and not the
+                    # signal that caused it)
+                    event_data = dict(
+                        zip(
+                            ("pid", "tgid", "exit_code", "exit_signal"),
+                            self._exit_proc_event.unpack(
+                                data[
+                                    self._base_proc_event.size : self._base_proc_event.size + self._exit_proc_event.size
+                                ]
+                            ),
+                        )
+                    )
+
+                    for callback in self._exit_callbacks:
+                        callback(event_data["pid"], event_data["tgid"], event_data["exit_code"])
+
+        # Cleanup
+        self._selector.unregister(self._socket)
+        self._selector.unregister(self._select_breaker_reader)
+        self._socket.close()
+        os.close(self._select_breaker)
+        os.close(self._select_breaker_reader)
+
+    @_raise_if_not_running
+    def stop(self):
+        self._should_stop = True
+        # Write to make select() return
+        os.write(self._select_breaker, b"\0")
+
+    @_raise_if_not_running
+    def register_exit_callback(self, callback: Callable):
+        self._exit_callbacks.append(callback)
+
+    @_raise_if_not_running
+    def unregister_exit_callback(self, callback: Callable):
+        self._exit_callbacks.remove(callback)
+
+
+_proc_events_listener = _ProcEventsListener()
```

Comment:
This `__init__` still runs some code (creates a socket, ...)

Let's push the initialization to `_ensure_thread_started` as well.